pokemongo_bot/cell_workers/evolve_pokemon.py

```diff
@@ -0,0 +1,152 @@
+            return False
+
+        # Evolve all is used - Use Lucky egg only at the first tick
+        if self.bot.tick_count is not 1 or not self.use_lucky_egg:
+            return True
+
+        lucky_egg_count = self.bot.item_inventory_count(Item.ITEM_LUCKY_EGG.value)
+
+        # Make sure the user has a lucky egg and skip if not
+        if lucky_egg_count > 0:
+            logger.log('Using lucky egg ... you have {}'.format(lucky_egg_count))
+            response_dict_lucky_egg = self.bot.use_lucky_egg()
+            if response_dict_lucky_egg:
+                result = response_dict_lucky_egg.get('responses', {}).get('USE_ITEM_XP_BOOST', {}).get('result', 0)
+                if result is 1:  # Request success
+                    logger.log('Successfully used lucky egg... ({} left!)'.format(lucky_egg_count - 1), 'green')
+                    return True
+                else:
+                    logger.log('Failed to use lucky egg!', 'red')
+                    return False
+        else:
+            # Skipping evolve so they aren't wasted
+            logger.log('No lucky eggs... skipping evolve!', 'yellow')
+            return False
+
+    def _get_candy_list(self, inventory_items):
+        candies = {}
+        for item in inventory_items:
+            candy = item.get('inventory_item_data', {}).get('candy', {})
+            family_id = candy.get('family_id', 0)
+            amount = candy.get('candy', 0)
+            if family_id > 0 and amount > 0:
+                family = self.bot.pokemon_list[family_id - 1]['Name'] + " candies"
+                candies[family] = amount
+
+        return candies
+
+    def _sort_and_filter(self, inventory_items):
+        pokemons = []
+        logic_to_function = {
+            'or': lambda pokemon: pokemon["cp"] >= self.evolve_above_cp or pokemon["iv"] >= self.evolve_above_iv,
+            'and': lambda pokemon: pokemon["cp"] >= self.evolve_above_cp and pokemon["iv"] >= self.evolve_above_iv
+        }
+        for item in inventory_items:
+            pokemon = item.get('inventory_item_data', {}).get('pokemon_data', {})
+            pokemon_num = int(pokemon.get('pokemon_id', 0)) - 1
+            next_evol = self.bot.pokemon_list[pokemon_num].get('Next Evolution Requirements', {})
+            pokemon = {
+                'id': pokemon.get('id', 0),
+                'num': pokemon_num,
+                'name': self.bot.pokemon_list[pokemon_num]['Name'],
+                'cp': pokemon.get('cp', 0),
+                'iv': self._compute_iv(pokemon),
+                'candies_family': next_evol.get('Name', ""),
+                'candies_amount': next_evol.get('Amount', 0)
+            }
+            if pokemon["id"] > 0 and pokemon["candies_amount"] > 0 and (logic_to_function[self.cp_iv_logic](pokemon)):
+                pokemons.append(pokemon)
+
+        if self.first_evolve_by == "cp":
+            pokemons.sort(key=lambda x: (x['num'], x["cp"], x["iv"]), reverse=True)
```

Comment:
@douglascamata 
The pokemon index I mentioned is the NUMBER you said.  We are actually the same understanding about the logic of those lines.  Let me describe more why I think sorting by pokemon number is not good.  As I mentioned in another thread. I have a patch to automatically evolve mid-tier pokemons like Ivysaur, Metapod or Pidgeotto which have further evolution. I tried to reserve candies for mid-tier pokemons with better potential (a.k.a better iv/cp) otherwise because the basic tier pokemon like Bulbasaur, Caterpie or Pidgey will be eariler to reach the candy requirement than their upper levels. In order to avoid starvation of the upper evolution, I rely on the sort, otherwise I might need to resort it again among those pokemons in the same family. If we sort by pokemon number, the higher level (which is mid-tier) of a family will get the chance to evolve. It's like a stronger first algorithm, not a better potential first algorithm. While if only sort by cp & iv, user have 2 options on choosing either stronger first or better potential first.   Furthermore, from the literal meaning of those config options of 'first_evolve_by', there is no pokemon index, right? I just think keeping the literal meaning with no surprise to the user is better.